src/bentoml/_internal/monitoring/base.py

```diff
@@ -0,0 +1,209 @@
+from __future__ import annotations
+
+import typing as t
+import logging
+import collections
+import contextvars
+
+MON_COLUMN_VAR: contextvars.ContextVar[
+    "dict[str, dict[str, str]] | None"
+] = contextvars.ContextVar("MON_COLUMN_VAR", default=None)
+MON_DATAS_VAR: contextvars.ContextVar[
+    "dict[str, collections.deque[t.Any]] | None"
+] = contextvars.ContextVar("MON_DATAS_VAR", default=None)
+
+DT = t.TypeVar("DT")
+MT = t.TypeVar("MT", bound="MonitorBase[t.Any]")
+
+MONITOR_REGISTRY: dict[str, type[MonitorBase[t.Any]]] = {}
+
+BENTOML_MONITOR_ROLES = {"feature", "prediction", "target"}
+BENTOML_MONITOR_TYPES = {"numerical", "categorical", "numerical_sequence"}
+
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorBase(t.Generic[DT]):
+    """
+    The base monitor class. All monitors should inherit from this class.
+    Subclasses should implement the following methods:
+    - export_schema
+    - export_data
+    to export the schema and data to the desired format.
+    """
+
+    PRESERVED_COLUMNS: tuple[str, ...] = ()
+
+    monitor_type: str | None
+
+    def __init_subclass__(cls, *, monitor_type: str | None = None):
```

Comment:
so that we can register it to the MONITOR_REGISTRY, which is used under the api.py

Since we expose the MonitorBase on monitoring, user can write their own like so

```python

class CustomMonitor(MonitorBase, monitor_type='my-monitor'): ...
```

Obv this is opt-in, as the previous behaviour should still work as expected.

Reply:
I think the previous version has the same effect, and easy to understand.